orderer/common/cluster/block_fetcher_test.go

```diff
@@ -0,0 +1,331 @@
+
+func TestBlockFetcherShuffleTimeOutDisable(t *testing.T) {
+	bf := cluster.BlockFetcher{}
+
+	bf.BlockSourceFactory = func(c cluster.FetcherConfig) cluster.BlockSource {
+
+		if c.Endpoints[0].Endpoint == "localhost:5100" {
+
+			bf.Logger.Infof("Block puller 0 Created")
+			bs := &mocks.BlockSource{}
+
+			bs.On("PullBlock", mock.Anything).Return(&common.Block{
+				Header:   &common.BlockHeader{Number: 1},
+				Data:     &common.BlockData{},
+				Metadata: &common.BlockMetadata{},
+			}).After(time.Second * 12)
+
+			bs.On("UpdateEndpoints", mock.Anything)
+			bs.On("Close", mock.Anything)
+			return bs
+		} else {
+			bf.Logger.Infof("Block puller 1 Created")
+			bs := &mocks.BlockSource{}
+			bs.On("PullBlock", mock.Anything).Return(&common.Block{
+				Header:   &common.BlockHeader{Number: 2},
+				Data:     &common.BlockData{},
+				Metadata: &common.BlockMetadata{},
+			}).After(time.Second * 12)
+
+			bs.On("UpdateEndpoints", mock.Anything)
+			bs.On("Close", mock.Anything)
+			return bs
+		}
+	}
+	// Disable ShuffleTimeout
+	bf.ShuffleTimeout = time.Duration(0)
+	bf.Logger = flogging.MustGetLogger("test")
+	bf.Config.Endpoints = []cluster.EndpointCriteria{{Endpoint: "localhost:5100"}, {Endpoint: "localhost:5200"}}
+
+	bf.Endpoints = []cluster.EndpointCriteria{{Endpoint: "localhost:5100"}, {Endpoint: "localhost:5200"}}
+
+	require.Equal(t, uint64(1), bf.PullBlock(1).Header.Number)
+	// source is not shuffled
+	require.Equal(t, uint64(1), bf.PullBlock(1).Header.Number)
+
+}
+
+func TestBlockFetcherNodeOffline(t *testing.T) {
+	bf := cluster.BlockFetcher{}
+
+	bf.BlockSourceFactory = func(c cluster.FetcherConfig) cluster.BlockSource {
+
+		if c.Endpoints[0].Endpoint == "localhost:5100" {
+
+			bf.Logger.Infof("Block puller 0 Created")
+			bs := &mocks.BlockSource{}
+
+			// return nil after 2 sec ( less than 50% of bf.Config.FetchTimeout)
+			// to test if the endpoint is offline
+			bs.On("PullBlock", mock.Anything).Return(nil).After(time.Second * 2)
+
+			bs.On("UpdateEndpoints", mock.Anything)
+			bs.On("Close", mock.Anything)
+			return bs
+		} else {
+			bf.Logger.Infof("Block puller 1 Created")
+			bs := &mocks.BlockSource{}
+			bs.On("PullBlock", mock.Anything).Return(&common.Block{
+				Header:   &common.BlockHeader{Number: 2},
+				Data:     &common.BlockData{},
+				Metadata: &common.BlockMetadata{},
+			}).After(time.Second * 2)
+
+			bs.On("UpdateEndpoints", mock.Anything)
+			bs.On("Close", mock.Anything)
+			return bs
+		}
+	}
+	// Disable ShuffleTimeout
+	bf.ShuffleTimeout = time.Duration(0)
+	bf.Logger = flogging.MustGetLogger("test")
+	bf.Config.Endpoints = []cluster.EndpointCriteria{{Endpoint: "localhost:5100"}, {Endpoint: "localhost:5200"}}
+	bf.Config.FetchTimeout = time.Duration(time.Second * 10)
+
+	bf.Endpoints = []cluster.EndpointCriteria{{Endpoint: "localhost:5100"}, {Endpoint: "localhost:5200"}}
+
+	require.Equal(t, uint64(2), bf.PullBlock(2).Header.Number)
+}
+
+func TestBlockFetcherBFTBehavior(t *testing.T) {
```

Comment:
can you please write the scenario you're trying to test?

Reply:
sure